design_drafts/controller_components.md

```diff
@@ -0,0 +1,48 @@
+# Loading Controller as Components
+
+## Components in a Nutshell
+
+ROS2 Eloquent introduced the concept of components.
+A component represents a node which can be conveniently loaded into a generic components container and so being executed without the node to compile the node in its own executable.
+Throughout the rest of this document, the word `node` is equivalent to `component`.
+The component container (a.k.a `ComponentManager`) takes ownership over the loaded components and can thus control their life cycle, in which components can be dynamically loaded and unloaded.
+That should be enough background for the scope of this document.
+For more details see https://index.ros.org/doc/ros2/Tutorials/Composition/
+
+## Controller Manager as a Component Manager
+
+The `ComponentManager` is in a sense very similar to the `ControllerManager` in which the controller manager holds the same logic of dynamically loading and unloading controllers.
+We therefore propose to leverage the functionality of the `rclcpp_components` within the controller manager, which provides all necessary services and command line tools needed to load components.
+
+### Type Safety of Components
+
+The component loading functionality differs from the `rclcpp_components` in a way that the controller manager has to make sure to load only components which adhere to the `ControllerInterface` and not load any arbitrary nodes.
+The current component container interface does not allow to filter for specific types and has to be modified accordingly.
+The proposed design lets the controller manager being a child class of the component manager and gives so the change to override the appropriate functions to guarantee a strict type safety.
```

Comment:
Is there an implication of this on the component manager side? Are the necessary functions properly exposed so overriding is possible?

Reply:
indeed, the component manager has to be made public and virtual. A PR for `rclcpp_components` can be made simultaneously with the one for the controller manager though.

Reply:
If there aren't virtual functions there yet, this will be an ABI-breaking change. Better check on it early ;) Could you please drop a note of this in the doc too, then I think we can close discussions & merge

Reply:
see https://github.com/ros2/rclcpp/issues/1010

Reply:
thanks! :+1: